File: lib/services/email.ts
import { Resend } from "resend";

import { env } from "@/lib/env";
import type { CurationPreviewItem, EmailPreview, ProfileRecord } from "@/lib/types";

const resend = env.RESEND_API_KEY ? new Resend(env.RESEND_API_KEY) : null;

export function buildEmailPreview(profile: ProfileRecord, items: CurationPreviewItem[]): EmailPreview {
  const dateLabel = new Intl.DateTimeFormat("en-US", { dateStyle: "long" }).format(new Date());
  const greeting = `Hey ${profile.name}, your AI desk is ready for ${dateLabel}.`;
  const introduction = items.length
    ? `The top stories today lean into ${items
        .slice(0, 3)
        .map((item) => item.sourceSlug.replace(/-/g, " "))
        .join(", ")}. The list below is ranked against your profile for practical, technically dense signal.`
    : "No ranked digests are available yet, so the system is holding a preview-only issue.";

  const markdown = [
    greeting,
    "",
    introduction,
    "",
    ...items.slice(0, profile.topN).flatMap((item) => [
      `## ${item.rank}. ${item.title}`,
      item.summary,
      `Source: ${item.sourceSlug}`,
      `[Read more](${item.url})`,
      ""
    ])
  ].join("\n");

  const html = `
    <div style="font-family:Arial,sans-serif;max-width:680px;margin:0 auto;padding:24px;color:#17222e;">
      <h2 style="font-size:24px;margin-bottom:12px;">${greeting}</h2>
      <p style="font-size:16px;line-height:1.6;">${introduction}</p>
      ${items
        .slice(0, profile.topN)
        .map(
          (item) => `
            <section style="padding:18px 0;border-top:1px solid #dce3ea;">
              <p style="font-size:12px;letter-spacing:0.08em;text-transform:uppercase;color:#8e3314;">#${item.rank} • ${item.sourceSlug}</p>
              <h3 style="margin:6px 0 8px;">${item.title}</h3>
              <p style="line-height:1.6;">${item.summary}</p>
              <a href="${item.url}" style="color:#8e3314;font-weight:700;text-decoration:none;">Open source</a>
            </section>
          `
        )
        .join("")}
    </div>
  `.trim();

  return {
    subject: `Daily AI News Desk | ${dateLabel}`,
    greeting,
    introduction,
    html,
    markdown,
    generatedAt: new Date().toISOString()
  };
}

export async function sendDigestEmail(preview: EmailPreview) {
  return sendDigestEmailToRecipient(preview, env.EMAIL_TO ?? null);
}

export async function sendDigestEmailToRecipient(preview: EmailPreview, recipientEmail: string | null) {
  if (!resend || !env.EMAIL_FROM || !recipientEmail) {
    return {
      sent: false,
      previewOnly: true,
      providerMessageId: null,
      error: !recipientEmail
        ? "No recipient email was available for this run."
        : !env.EMAIL_FROM
          ? "EMAIL_FROM is not configured."
          : "RESEND_API_KEY is not configured."
    };
  }

  try {
    const result = await resend.emails.send({
      from: env.EMAIL_FROM,
      to: recipientEmail,
      subject: preview.subject,
      html: preview.html,
      text: preview.markdown
    });

    if (result.error) {
      return {
        sent: false,
        previewOnly: false,
        providerMessageId: null,
        error: result.error.message ?? "Resend rejected the email."
      };
    }

    return {
      sent: true,
      previewOnly: false,
      providerMessageId: result.data?.id ?? null,
      error: null
    };
  } catch (error) {
    return {
      sent: false,
      previewOnly: false,
      providerMessageId: null,
      error: error instanceof Error ? error.message : "Unknown email send failure."
    };
  }
}

File: app/services/email.py
import html
import os

import markdown
import resend
from dotenv import load_dotenv

load_dotenv()

EMAIL_FROM = os.getenv("EMAIL_FROM")
EMAIL_TO = os.getenv("EMAIL_TO")
RESEND_API_KEY = os.getenv("RESEND_API_KEY")


def send_email(subject: str, body_text: str, body_html: str = None, recipients: list = None):
    if not RESEND_API_KEY:
        raise ValueError("RESEND_API_KEY environment variable is not set")
    if not EMAIL_FROM:
        raise ValueError("EMAIL_FROM environment variable is not set")

    if recipients is None:
        if not EMAIL_TO:
            raise ValueError("EMAIL_TO environment variable is not set")
        recipients = [EMAIL_TO]

    recipients = [r for r in recipients if r is not None]
    if not recipients:
        raise ValueError("No valid recipients provided")

    resend.api_key = RESEND_API_KEY

    params: resend.Emails.SendParams = {
        "from": EMAIL_FROM,
        "to": recipients,
        "subject": subject,
        "text": body_text,
    }
    if body_html:
        params["html"] = body_html

    resend.Emails.send(params)


def markdown_to_html(markdown_text: str) -> str:
    rendered_html = markdown.markdown(markdown_text, extensions=["extra", "nl2br"])
    return f"""<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <style>
        body {{
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
            line-height: 1.6;
            color: #333;
            max-width: 600px;
            margin: 0 auto;
            padding: 20px;
            background-color: #ffffff;
        }}
        h2 {{
            font-size: 18px;
            font-weight: 600;
            color: #1a1a1a;
            margin-top: 24px;
            margin-bottom: 8px;
            line-height: 1.4;
        }}
        h3 {{
            font-size: 16px;
            font-weight: 600;
            color: #1a1a1a;
            margin-top: 20px;
            margin-bottom: 8px;
            line-height: 1.4;
        }}
        p {{
            margin: 8px 0;
            color: #4a4a4a;
        }}
        strong {{
            font-weight: 600;
            color: #1a1a1a;
        }}
        em {{
            font-style: italic;
            color: #666;
        }}
        a {{
            color: #0066cc;
            text-decoration: none;
            font-weight: 500;
        }}
        a:hover {{
            text-decoration: underline;
        }}
        hr {{
            border: none;
            border-top: 1px solid #e5e5e5;
            margin: 20px 0;
        }}
        .greeting {{
            font-size: 16px;
            font-weight: 500;
            color: #1a1a1a;
            margin-bottom: 12px;
        }}
        .introduction {{
            color: #4a4a4a;
            margin-bottom: 20px;
        }}
        .article-link {{
            display: inline-block;
            margin-top: 8px;
            color: #0066cc;
            font-size: 14px;
        }}
    </style>
</head>
<body>
{rendered_html}
</body>
</html>"""


def digest_to_html(digest_response) -> str:
    from app.agent.email_agent import EmailDigestResponse

    if not isinstance(digest_response, EmailDigestResponse):
        if hasattr(digest_response, "to_markdown"):
            return markdown_to_html(digest_response.to_markdown())
        return markdown_to_html(str(digest_response))

    html_parts = []
    greeting_html = markdown.markdown(
        digest_response.introduction.greeting, extensions=["extra", "nl2br"]
    )
    introduction_html = markdown.markdown(
        digest_response.introduction.introduction, extensions=["extra", "nl2br"]
    )
    html_parts.append(f'<div class="greeting">{greeting_html}</div>')
    html_parts.append(f'<div class="introduction">{introduction_html}</div>')
    html_parts.append("<hr>")

    for article in digest_response.articles:
        html_parts.append(f"<h3>{html.escape(article.title)}</h3>")
        summary_html = markdown.markdown(article.summary, extensions=["extra", "nl2br"])
        html_parts.append(f"<div>{summary_html}</div>")
        html_parts.append(
            f'<p><a href="{html.escape(article.url)}" class="article-link">Read more -&gt;</a></p>'
        )
        html_parts.append("<hr>")

    html_content = "\n".join(html_parts)

    return f"""<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <style>
        body {{
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
            line-height: 1.6;
            color: #333;
            max-width: 600px;
            margin: 0 auto;
            padding: 20px;
            background-color: #ffffff;
        }}
        h3 {{
            font-size: 16px;
            font-weight: 600;
            color: #1a1a1a;
            margin-top: 20px;
            margin-bottom: 8px;
            line-height: 1.4;
        }}
        p {{
            margin: 8px 0;
            color: #4a4a4a;
        }}
        strong {{
            font-weight: 600;
            color: #1a1a1a;
        }}
        em {{
            font-style: italic;
            color: #666;
        }}
        a {{
            color: #0066cc;
            text-decoration: none;
            font-weight: 500;
        }}
        a:hover {{
            text-decoration: underline;
        }}
        hr {{
            border: none;
            border-top: 1px solid #e5e5e5;
            margin: 20px 0;
        }}
        .greeting {{
            font-size: 16px;
            font-weight: 500;
            color: #1a1a1a;
            margin-bottom: 12px;
        }}
        .introduction {{
            color: #4a4a4a;
            margin-bottom: 20px;
        }}
        .article-link {{
            display: inline-block;
            margin-top: 8px;
            color: #0066cc;
            font-size: 14px;
        }}
        .greeting p {{
            margin: 0;
        }}
        .introduction p {{
            margin: 0;
        }}
        div {{
            margin: 8px 0;
            color: #4a4a4a;
        }}
        div p {{
            margin: 4px 0;
        }}
    </style>
</head>
<body>
{html_content}
</body>
</html>"""


def send_email_to_self(subject: str, body: str):
    if not EMAIL_TO:
        raise ValueError("EMAIL_TO environment variable is not set. Please set it in your .env file.")
    send_email(subject, body, recipients=[EMAIL_TO])


if __name__ == "__main__":
    send_email_to_self("Test from Python", "Hello from my script.")
